nservicebus/messaging/evolving-contracts.md

```diff
@@ -1,108 +1,83 @@
-
-At that point it makes more sense to break down the contracts into multiple assemblies. A possible structure solution is to have multiple assemblies per endpoint with a naming convention similar to:
-
-* EndpointName.Commands
-* EndpointName.Events
-
-Or even more strict:
-
-* EndpointName.AutonomousComponent.Commands
-* EndpointName.AutonomousComponent.Events
-
-While it might make sense to directly reference the \*.Commands assemblies from the sender and receiver endpoints, it might not for the \*.Events assemblies as they are used by many subscribing endpoints. In those cases, considering sharing the assembly through NuGet packages as opposed to using a direct reference.
+---
+title: Evolving Message Contracts
+summary: Guidelines for choosing a strategy for evolving message contracts
+reviewed: 2022-02-10
+component: Core
+isLearningPath: true
+---
+
+In message-based systems, the messages are part of a contract, which defines how services communicate with each other.
+
+
+Evolving contracts over time is challenging and an appropriate strategy should be reviewed and customized for each system. When evolving message contracts, consider the following:
+
+
+* Endpoints updated to the latest message contract might still receive messages using the old contract. Senders might still use the old contract, or not all in-flight messages (messages waiting to be consumed in input queues) have been processed before the upgrade.
+
+* Endpoints updated to the latest message contract might send messages, using the new contract, to endpoints still based on the old contract version.
+
+
+Generally, the problem can't be resolved at the infrastructure level; therefore, NServiceBus users must analyze their systems, consider how they are expected to evolve, and define the strategy which will make the most sense in their particular circumstances.
+
+This article presents basic guidelines for choosing a contract evolution strategy, avoiding common mistakes, and ensuring that contracts will be easy to evolve over time.
+
+
+TODO: link to new document
+Note: Ensure that message contracts follow the general [messages design guidelines](/nservicebus/messaging/messages-events-commands.md#designing-messages).
+
+## Adding new message contracts
+
+When adding new message contracts to a contracts assembly, update the endpoints receiving or subscribing to the new contract type first. If the sender/publisher endpoints are updated first, receivers and subscribers will not process messages for the new contract type.

```

Comment:
* Events: usually update publisher first and then subscribers can subscribe at any time. (if there is a requirement to receive all events, subscriber should be updated first)
* Commands: Upgrade the receiver first, otherwise the receiver would have "handler not found" exceptions

Reply:
this approach can be used for completely new messages but also for "new versions" that are just published twice.